src/asm/lexer.c

```diff
@@ -156,28 +156,71 @@ YY_BUFFER_STATE yy_create_buffer(FILE *f)
 	if (pBuffer == NULL)
 		fatalerror("%s: Out of memory!", __func__);
 
-	uint32_t size;
+	size_t size = 0, capacity = -1;
+	char *buf = NULL;
 
-	fseek(f, 0, SEEK_END);
-	size = ftell(f);
-	fseek(f, 0, SEEK_SET);
-
-	/* Give extra room for 2 newlines and terminator */
-	uint32_t capacity = size + 3;
+	/*
+	 * Check if we can get the file size without implementation-defined
+	 * behavior:
+	 *
+	 * From ftell(3p):
+	 * [On error], ftell() and ftello() shall return −1, and set errno to
+	 * indicate the error.
+	 *
+	 * The ftell() and ftello() functions shall fail if: [...]
+	 * ESPIPE The file descriptor underlying stream is associated with a
+	 * pipe, FIFO, or socket.
+	 *
+	 * From fseek(3p):
+	 * The behavior of fseek() on devices which are incapable of seeking
+	 * is implementation-defined.
+	 */
+	if (ftell(f) != -1) {
+		fseek(f, 0, SEEK_END);
+		capacity = ftell(f);
+		rewind(f);
+	}
 
-	pBuffer->pBufferRealStart = malloc(capacity + SAFETYMARGIN);
+	// If ftell errored or the block above wasn't executed
+	if (capacity == -1)
+		capacity = 4096;
+	// Handle 0-byte files gracefully
+	else if (capacity == 0)
+		capacity = 1;
+
+	while (!feof(f)) {
+		if (buf == NULL || size >= capacity) {
+			if (buf)
+				capacity *= 2;
+			/* Give extra room for 2 newlines and terminator */
+			buf = realloc(buf, capacity + SAFETYMARGIN + 3);
+
+			if (buf == NULL)
+				fatalerror("%s: Out of memory for buffer!",
+					   __func__);
+		}
 
-	if (pBuffer->pBufferRealStart == NULL)
-		fatalerror("%s: Out of memory for buffer!", __func__);
+		char *bufpos = buf + SAFETYMARGIN + size;
+		size_t read_count = fread(bufpos, 1, capacity - size, f);
 
-	pBuffer->pBufferStart = pBuffer->pBufferRealStart + SAFETYMARGIN;
-	pBuffer->pBuffer = pBuffer->pBufferRealStart + SAFETYMARGIN;
+		if (read_count == 0 && !feof(f))
+			fatalerror("%s: fread error", __func__);
 
-	size = fread(pBuffer->pBuffer, sizeof(uint8_t), size, f);
+		size += read_count;
+	}
 
+	pBuffer->pBufferRealStart = buf;
+	pBuffer->pBufferStart = buf + SAFETYMARGIN;
+	pBuffer->pBuffer = buf + SAFETYMARGIN;
 	pBuffer->pBuffer[size] = 0;
 	pBuffer->nBufferSize = size;
 
+	/* This is added here to make the buffer scaling above easy to express,
+	 * while taking the newline space into account
+	 * for the `yy_buffer_append`s below.
+	 */
```

Comment:
I think you've tried to make the old "Give extra room for 2 newlines and terminator" comment a little clearer, while also taking away the reasoning for this magic number… 🙂 If you're happy to reword this, to make it clear here that the later `yy_buffer_append`s might add two newlines and a terminator (justifying the `3` here), that'd be great!